docs/proposals/fluentd-limit-requests-to-apiserver.md

```diff
@@ -0,0 +1,63 @@
+# Modifications to Fluentd Metadata Retrieval Proposal
+
+##Motivation
+In the current design (v3.4) of the EFK stack, there is a Fluentd pod deployed to each Openshift node. Fluentd is responsible for reading container logs, adding pod metadata (e.g. labels, namespace) to individual log entries, and sending the logs to the Elasticsearch cluster. The pod metadata is sourced by watching the Openshift API server for Pod information via the [Kubernetes Metadata Filter](https://github.com/fabric8io/fluent-plugin-kubernetes_metadata_filter)[2]. Testing at larger scales of Openshift clusters has shown issues where Fluentd is unable to connect to the API server because it is limited by the **maxRequestsInFlight** configuration setting [1].  This results in logs not being shipped to the aggregator and thus being unavailable.
+
+##Proposal - Short Term (3.4)
+1. Leave the value as-is and note that it must be adjusted to account for # of nodes plus some headroom (swag @ 10%?).
+1. Increase the default to our current maximum supported node count, plus some headroom.
+1. Increasing maxRequestsInflight upstream
+
+##Proposal - Long Term
+###Option 1 - Admission Controller
+1. Create a simple admission controller which simply embeds the necessary meta-data as env-variables into the pods on submission.  In this way every pod that is run on the system will have its data locally.
+1. Mount the docker socket into the Fluentd pod such that it can query containers env-variables.
+1. Remove any watches from the Fluentd pods, thereby eliminating excessive traffic from the node.
+
+This will require:
+* Admission controller introduced to [Openshift Origin](https://github.com/openshift/origin)
+* Deployment changes (e.g. deployer or ansible)
+* A new Fluentd plugin
+* Support setup (e.g. ci, packaging, github/enterprise repo)
+
+Questions:
+1. How long is metadata available?  It is limited by Pod GC?
+1. Can we completely eliminate a filter[1] to provide metadata? Consider keeping some implementation that contacts API server as a fallback mechanism
+
+###Option 2 - Rsyslog collector to Fluentd (Augmentor)
+Utilize rsyslog to collect logs and forward them to Fluentd.  There is initial work to support this and would allow a 'vanilla' rsyslog implementation to work in conjunction with some smaller number of Fluentd pods to add metadata.
+
+This will require:
+* Deployment changes (e.g. deployer or ansible)
+* Packing and containerization of rsyslog?
+
+Questions:
+1. Does this require any Fluentd plugin changes?
+1. Does this require modifications to Fluentd? (e.g. how it receives logs, ships logs)
+1. How many Fluentd instance are needed?
+1. Where do Fluentd instance get deployed? Infra nodes only?
+
+###Option 3 - Fluentd (Collector) to Fluentd (Augmentor)
+This is similiar to **Option 2** but it utilizes Fluentd as the collector instead of rsyslog.
+
+This will require:
+* Deployment Changes (e.g. deployer or ansible)
+
+Questions:
+See questions from **Option 2**
```

Comment:
This also becomes complicated to scale up and deploy for the reasons that we would have two different configmaps to maintain, one that simply gathers and one that enriches the data.

They would need to be scaled independently so that's potentially two daemonsets.

We could then have some nodes where there are two Fluentd instances deployed, one of which would be more resource intensive (cpu, network, memory).